[KEP-2941]: Supporting Dynamic Resource Allocations in Kueue (#5149)
kubernetes-sigs/kueue: keps/2941-DRA/README.md

  - [Using ResourceFlavor for DeviceClass Mapping](#using-resourceflavor-for-deviceclass-mapping)
<!-- /toc -->

## Summary

Dynamic Resource Allocation (DRA) is a major effort to improve device support in Kubernetes.
It changes how one can request resources in a myriad of ways.

## Motivation

Dynamic Resource Allocation (DRA) provides the groundwork for more sophisticated device allocations to Pods.
Quota management is about enforcing rules around the use of resources.
For example, GPUs are resource constrained and a popular request is the ability to enforce fair sharing of GPU resources.
With these devices, many users want access and sometimes some users want the ability to preempt other users if their
workloads have a higher priority. Kueue provides support for this.

DRA provides a future where users could schedule partitionable GPU devices (MIG) or time slicing. As devices gain a 
more robust way to schedule, it is important to walk through how support of DRA will work with Kueue.

### Background

DRA has four APIs that are relevant for a Kueue:

- ResourceClaims
- ResourceClaimTemplates
- DeviceClasses

> - ResourceSlices

Reviewer: Yes, I think we need to scope for the first iteration of alpha, otherwise we risk not delivering in 0.13

    app: job
    kueue.x-k8s.io/queue-name: user-queue
spec:
  template:
    spec:
      restartPolicy: Never
      containers:
      - name: ctr0
        image: ubuntu:22.04
        command: ["bash", "-c"]
        args: ["export; sleep 9999"]
        resources:
          claims:
          - name: gpu
          requests:
            cpu: 1
            memory: "200Mi"
      resourceClaims:
      - name: gpu
        resourceClaimTemplateName: single-gpu
```

#### Example Driver Cluster Resources

The dra-example-driver creates a ResourceSlice for each node and a DeviceClass named `gpu.example.com` for the entire 
cluster.

##### ResourceSlices

ResourceSlices are meant for communication between drivers and the control planes. These are not expected to be used for
workloads.

##### DeviceClasses

Each driver creates a device class and every resource claim will reference the device class. The dra-example-driver has
a simple device class named `gpu.example.com`. This will be the way to enforce quota limits.

### Goals

- Users can submit workloads using resource claims and Kueue can monitor the usage.
- Admins can enforce the quota for number of devices for a given DeviceClass.

> ### Non-Goals

Reviewer 2: Additionally, I'm suspecting the first Kueue with the DRA feature does not have compatibility for the MultiKueue.
Because the ResourceClaimTemplate and ResourceClaim responsibilities are out of Kueue.

In other words, batch users need access to all clusters and create a ResourceClaimTemplate in their namespaces before they submit Jobs.
However, in the dynamic cluster creation / deletion environment, it's challenging, I believe.

In the future DRA support, I think we need to consider how to support it for MultiKueue the same as TopologyAwareScheduling.

cc @mimowo

Reviewer: Actually, I think MultiKueue may work ootb if users repeat the configuration in management and worker cluster. 
I'm ok with a note as non-goal, but maybe just a graduation point for beta is enough.

Reviewer 2: Let's add MultiKueue to the Beta criteria.

Author: added.

- We are limiting scope for DRA to structured parameters (beta in 1.32 and 1.33)
  - Support for alpha features like DRADeviceTaints, DRAAdminAccess, DRAPrioritizedLists and DRAPartitionableDevices
    will not be included.
- This design does not work with Topology Aware Scheduling feature of Kueue. It is a significant amount of work, will be
   addressed in the future with a separate body of work

## Proposal

This proposal is to extend the APIs for allowing workloads using DRA APIs to be tallied against quota management, 
borrowing and preemptable scheduling. This includes modifying:

> 1. The `ResourceFlavor` API to allow creating a mapping canonical name -> device class names list

Reviewer: ```suggestion
1. Introduce API to allow creating a mapping canonical name -> device class names list
```
since we migrate to use CRD for the mapping.

In order to mitigate this risk, Kueue can take the following approach:
1. For DRAPrioritizedLists: all the mentioned device classes in the request will be counted against the quota
2. For DRAAdminAccess: This feature can only be enabled in admin namespace, therefore it should be skipped for being 
   counted against ClusterQuota. This is a different stance than Kubernetes ResourceQuota because kubernetes
   ResourceQuota is namespace scoped. As a result, admin users can account for quota independent of user workloads.
   On the contrary, with Kueue, since quota is part of ClusterQuota a cluster scoped object, admin workloads using
   devices in admin namespace if counted against quota, will eat up quota meant for user workloads.
3. For ResourceClaims with allocation mode `All`: worst-case scenario of the max number of devices that could be
   allocated to a single claim will be used against quota.

## Design Details

A new feature gate DynamicResourceAllocation will be introduced, allowing users to test it in dev environments while
making the changes dormant for production users. The following sections will explain the design in detail.

> ### DeviceClassMapping API

Reviewer: I would suggest naming the CRD more generically like "DynamicResourceAllocation" so that we have natural place to expand the configuration in the future.

We also discussed this with @tenzen-y recently offline

Author: > I would suggest naming the CRD more generically like "DynamicResourceAllocation" so that we have natural place to expand the configuration in the future.

With this name my concern is that we are tying the kueue abstractions too closely with the k8s feature. In essence, what are introducing is a Kueue abstraction, a ResourceMapping, that could be extended in the future for example for non-dra use-cases, say the ResourceTransformations with device-plugins could be housed here in this API.

For that reason, I wanted to name this CRD generically, how about `ResourceMapping`? However I am open to be convinced otherwise.

Reviewer: In any case since this is going to be Alpha, then we can relatively easily change our minds in that phase

Reviewer: @alaypatel07 

TLDR: I had a chance to discuss this with @tenzen-y, and we agreed to go with DynamicResourceAllocation CRD in alpha. 

The main reason for using CRD instead of configmap is that configmap is already beta, so we would need to be more careful about API changes. We expect DRA to evolve in the coming releases of k8s, so it is preferred to go via Alpha in Kueue.

Regarding the generic "ResourcesConfigurations" CRD we prefer specific, because sometimes the boundaries between features and fields are fuzzy. For example, MultiKueue may want to introduce some specific resource names, and it would not be clear where to declare the parameters.

Going forward, we will consider moving the configmap to CRD https://github.com/kubernetes-sigs/kueue/issues/5646. If this happens before moving DRA to Beta, then we could merge the configs. If this does not happen by then, then we may end up with two different configuration CRDs (global and DRA), we are ok with both possibilities.

Author: @mimowo de-coupling current configmap API and DRA and having a possibility to merge configs in bet makes a lot of sense to me, thanks for the summary. I will re-work the design to name this CRD DynamicResourceAllocation. Although I have one small concern, DynamicResourceAllocation is a K8s concept, the CRD we are introducing is a kueue concept, are the abstractions little ambiguous for users?

Reviewer: I think this is ok, because the API is inside the Kueue Group, and the instance will be inside the kueue-system namespace. 

Also, I think it is actually beneficial for discover ability, if we name it say ResourceMapping users may not know immediately this is for DRA configuration. Moreover the name may not fit some sub features of DRA which are not related directly to resource mapping.

Reviewer 2: @mimowo Thank you for adding our discussion result and thank you @alaypatel07 for reflecting my opinion to this KEP.

Author: marking this Resolved.

```golang

> // DeviceClassMapping is a cluster-scoped singleton CRD that maps device classes to canonical names
> // used in ClusterQueue resource quotas.

Reviewer: ```suggestion
// DeviceClassMapping is a cluster-scoped singleton CRD that maps device classes to canonical names
// used in ClusterQueue resource quotas. It is singleton as "default" is the only allowed named for the
// CRD instance.
```

type DynamicResourceAllocationConfig struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec defines the desired state of DynamicResourceAllocationConfig
    Spec DynamicResourceAllocationConfigSpec `json:"spec"`
}

// DynamicResourceAllocationConfigSpec defines the mappings between device classes and canonical names
type DynamicResourceAllocationConfigSpec struct {
    // DynamicResources is a list of mappings from canonical names to device classes
    // +listType=atomic

>     DynamicResources []DeviceClassMapping `json:"dynamicResources"`

Reviewer 2: @alaypatel07 could you propose the assuming number of dynamicResources?
Is 8 enough? Or do you have stories that need more and more resources?

>     DynamicResources []DeviceClassMapping `json:"dynamicResources"`

Author: https://github.com/kubernetes-sigs/kueue/pull/5149/files#r2163105420

we agreed to starting with 8 and re-consider in beta.

}

// DeviceMapping defines a mapping from a canonical name to a list of device classes
type DeviceMapping struct {
    // Name is the canonical name that will be referred to in ClusterQueue and Workload status

>     Name corev1.ResourceName `json:"name"`

Reviewer: No strong view, but maybe we call it "CanonicalName" since we use it in the doc in many places, and when talking?

Author: I have considered something like this, but it has the following problem:
```yaml
apiVersion: kueue.x-k8s.io/v1alpha1
kind: DeviceClassMapping
metadata:
  name: "default" 
spec:
  dynamicResources:
  - canonnicalName: gpus
    deviceClassNames:
    - gpu.example.com
---
apiVersion: kueue.x-k8s.io/v1beta1
kind: ClusterQueue
metadata:
  name: "gpus-cluster-queue"
spec:
  resourceGroups:
  - coveredResources: ["cpu", "memory", "gpus"]
    flavors:
    - name: "default-gpu-flavor"
      resources:
      - name: "gpus"                                                     # <--- the fields here do not match
        nominalQuota: 9
```

From UX point of view, keeping the field names consistent when cross-referencing in clusterqueue seem the most natural, DRA feature in k8s does this in its APIs.

Reviewer 3: I tend to agree, I like the symmetry of having both fields called `name`.

Reviewer: got it, I'm ok as is, just poining out as a potential thing to consider for @tenzen-y too

Reviewer: This is Alpha so I dont think we need to complicate the API for unknowns. If we discover the new features require different organization we can add it later. I suggest to focus on the current scope and make it easy to use.

My clear preference is option a) from Yuki which is quite clear from the CRD name, so I think no need to repeat terms like dynamic.

Reviewer 2: I just confirmed if breaking APIs are acceptable for you since someone sometimes requests backward compatibility even for alpha API, even for beta API. Actually, you requested to avoid cq (beta1 -> beta2) and cohort (alpha1 -> beta1) field name changes previously.

As long as we follow the kubernetes API graduation steps, any API change should be acceptable.
If anyone request such compabiilities for alpha and beta API, the Alpha API should be (a).
However, if someone potentially requests backward compabitilities, we should introduce more composable APIs in the alpha level.

Again, this is just confirmation if API break is acceptable even if it change just field name.

Reviewer 2: @alaypatel07 Thank you for summarizing that. In this KEP, we can take (a) approach in that case.

I guess that PrioritizedList and ConsumableCapacity are the biggest ones for Kueue since PrioritizedList is similar to Flavor fallback mechanism and ConsumableCapacity increase scheduling accuraccy, IIUC.

For PrioritizedList, if we support it, we will have a 2-level fallback mechanism, 1 is ResourceFlavor level, 2 is ResourceClaim level. I think we should discuss how to mitigate increase of complexity and entropy when we support PrioritizedList since it will generate n × m patterns in case of fallback where n is the number of ResourceFlavors in cq and m is the number of device requests in ResourceClaim (Template).

For ConsumableCapacity, I guess that this is needed for TAS, which means TAS should consider ConsumableCapacities in addition to Node allocatable resources when it simulates scheduling.
When TAS is disabled, I think we can ignore the ConsumableCapacity because the cq quota is just virtual. So Kueue don't care actual resource availabilities in case of non TAS mode.

One unclear point is whether or not cluster-autoscaler can understand ConsumableCapacity when it scale node. This is a bit of an important problem for PrivisioningRequest support in Kueue, IIUC. Because if CA can not recognize the ConsumableCapacity, it might scale out small nodes rather than the expected resources.

In any case, I'm open to discussion for DRA support in Kueue.

Author: I think all of these are good follow-up discussions for beta.  

To me it seems these changes could be additive to the DynamicResourceAllocationConfig struct instead of "breaking".  So as decided earlier, I have updated the design with option (a)

    // DeviceClassNames lists the names of all the device classes that will count against
    // the quota defined for this canonical name
    // +listType=atomic

>     DeviceClassNames []corev1.ResourceName `json:"deviceClassNames"`

Reviewer 2: Could you add limitation for the number of deviceClassNames?

Reviewer 2: 8 sounds good to me as well, thanks.

}
```

The ResourceFlavor spec will have a field called dynamicResource. The cluster admin has to define the list of
deviceClasses that will be represented by a canonical name and can be used to define quotas in cluster queue. With this
a ClusterQueue with nominalQuota can be defined as follows:

```yaml

> apiVersion: kueue.x-k8s.io/v1beta1

Reviewer: I would suggest to start with "Alpha" as the API is expected to grow.

kind: DeviceClassMapping
metadata:

>   name: "cluster-device-mapping"  # Fixed name - singleton CR, only one instance allowed

Reviewer: ```suggestion
  name: "default"  # Fixed name - singleton CR, only one instance allowed
```
for simplicity

>   name: "default"  # Fixed name - singleton CR, only one instance allowed

Reviewer: I would probably suggest keeping it under kueue-system namespace. This way it is closer to the ConfigMap model, which is intrinsically namespaced. Maybe using helm one could imagine two separete installations of Kueue. It is very unlikely, but possible.

Reviewer 2: As I posted in https://github.com/kubernetes-sigs/kueue/pull/5149#discussion_r2161830900, I'm leaning toward to (3). The cluster admin can select which cq can manage which namespaces based on namespaceSelector in cq.

After we introduce TAS to kueue, kueue-controller-manager retrieves all node information as a dedicated cache separate from the informer cache, which increases the memory footprint and kueue scheduling duration.

The sharding kueue-controller-manager concept would mitigate this problem.

    flavors:
    - name: zone-a-flavor
      resources:
      - name: whole-gpus
        nominalQuota: 2
    - name: zone-b-flavor
      resources:
      - name: fast-gpus
        nominalQuota: 3
```

When a device class like `gpus.example.com` maps to multiple canonical names (`whole-gpus` and `fast-gpus`), Kueue must
determine which quota to use for admission.

Kueue resolves this ambiguity using a top-down flavor matching process:

1. Examine flavors in the order they are listed in the ClusterQueue spec
2. For each flavor, check if any of its resources match the canonical names that correspond to the requested device class
3. Select the first flavor that contains a resource mapping to the device class and has sufficient quota available

For example, when a workload requests `gpus.example.com`:
- First check `zone-a-flavor` for available `whole-gpus` quota
- If `zone-a-flavor` is exhausted, check `zone-b-flavor` for available `fast-gpus` quota  
- If no flavors have available quota, the workload waits in queue

This approach ensures predictable behavior where the ClusterQueue flavor ordering determines quota evaluation priority.

> ### Workloads

Reviewer 4: @pohly mentioned to me that these resources are immutable for now.

> ### Workloads

Reviewer 2: Thank you for letting me know.
@alaypatel07, could you mention ResourceClaim immutabilities in the Notes section?

> ### Workloads

Author: +1 yes they are immutable, as part of the discussions in this design we double checked the latest APIs. I'll document it.

When a user submits a workload and DynamicResourceAllocation feature gate is on, Kueue will do the following:

> 1. Claims will be read from resources.claims in the PodTemplateSpec.

Author: +1 will be good to clarify. Out of curiosity, does DRA plan on support claims in pod level resources?

> 1. Claims will be read from resources.claims in the PodTemplateSpec.

Author: thinking about this more, Kueue doesnt care if the claims are referenced in container level resources or pod level within the podSpec. As per this design, Kueue will be looking at pod.spec.resourceClaims to count it against the quota.

> 1. Claims will be read from resources.claims in the PodTemplateSpec.

Author: updated.

> 1. Claims will be read from resources.claims in the PodTemplateSpec.

Reviewer 5: Just a pedantic nit that pod-level resources will be beta in 1.34. Though I concur that there are is no mention of resource claim (or DRA generally) in the updated pod-level resources KEP that was updated today for 1.34:

https://github.com/kubernetes/enhancements/pull/5362

cc @ndixita @pohly @johnbelamaric

2. ResourceClaimSpec will be looked up either by using:
   1. the name of the ResourceClaimTemplate or
   2. the name of the ResourceClaim

   Both ResourceClaimTemplate or ResourceClaim will be in the same namespace as the workload.
3. From the ResourceClaimSpec, the deviceClassName will be read.
4. Every claim, deviceClassName for each request will be looked at
   1. For the workload a deviceClassMap will be created by:
      1. Retrieving the singleton DynamicResourceAllocationConfig resource (named "default")
      2. Creating a map from DeviceClass names to canonical resource names using the
          DynamicResourceAllocationConfig.spec.dynamicResources entries
      3. For resource claims that are already allocated, Kueue will maintain a mapping called `admittedResourceClaims`

>          where the key follows the format `{clusterQueueName}/{resourceClaimNamespace}/{resourceClaimName}`

Reviewer: nit: this format of the key sounds like a technical detail. Actually in the cache.go we already have clustequeue caches so we could just keep the key per CQ cache. So, I consider it a technical detail which may change as we progress with the implementation.

in the form of:
- <package>: <date> - <current test coverage>

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

TBD
- `<package>`: `<date>` - `<test coverage>`

#### Integration tests

Integration tests in Kueue use controller-runtime's envtest framework, which provides a real Kubernetes API server
without requiring kubelet or other cluster components. While DRA device allocation requires kubelet plugins, the core
DRA integration functionality for Kueue can be tested at the integration level by:

- Testing DeviceClassMapping CRD creation and validation
- Verifying workload admission logic with DRA resource claims
- Testing quota enforcement against device class mappings
- Validating resource counting and flavor assignment for DRA resources

The integration tests would focus on Kueue's quota management and admission logic rather than actual device allocation,
using mock ResourceClaims and DeviceClasses to simulate DRA workloads.

#### E2E Test

Use existing dra-example-driver or Kubernetes test driver for e2e testing.

### Graduation Criteria

#### Alpha

> - the implementation behind the [feature gate](add link to the section) flag in alpha

Reviewer: ```suggestion
- the implementation behind the feature gate flag in alpha
```

   allocated to a single claim will be used against quota.

## Design Details

A new feature gate DynamicResourceAllocation will be introduced, allowing users to test it in dev environments while
making the changes dormant for production users. The following sections will explain the design in detail.

### DynamicResourceAllocationConfig API

```golang
// DynamicResourceAllocationConfig is a singleton CRD that maps resource names todevice classes
// used in ClusterQueue resource quotas. It is singleton as "default" is the only allowed named for the CRD instance in
// Kueue namespace.
type DynamicResourceAllocationConfig struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec defines the desired state of DynamicResourceAllocationConfig
    Spec DynamicResourceAllocationConfigSpec `json:"spec"`
}

// DynamicResourceAllocationConfigSpec defines the mappings between resource names and device classes
type DynamicResourceAllocationConfigSpec struct {
    // Resources is a list of mappings from resource name to device classes
    // +listType=atomic
    // +kubebuilder:validation:MaxItems=16

>     Resources []Resources `json:"resources"`

Reviewer: ```suggestion
    Resources []DynamicResources `json:"resources"`
```
or DRAResources

Author: https://github.com/kubernetes-sigs/kueue/pull/5149#discussion_r2161883129

I thought @tenzen-y wanted alignment in field name and struct name.

Reviewer: https://github.com/kubernetes-sigs/kueue/pull/5149#discussion_r2161883129

actually suggests DanamIcResource as the struct name, which is even better. 

Generally struct names can be more specific as they are global. Field names can be shorter and rely on the context.

Reviewer 2: > https://github.com/kubernetes-sigs/kueue/pull/5149#discussion_r2161883129
> 
> I thought @tenzen-y wanted alignment in field name and struct name.

When I left the comment, the field name was different. So, the comment has already been outdated.
As @mimowo mentioned, we want to use the API specific name as a Go API prefix to avoid unexpected conflicts with other object.

}

// Resources defines a mapping from a resource name to a list of device classes
type Resources struct {
    // Name is the resource name that will be referred to in ClusterQueue and Workload status
    Name corev1.ResourceName `json:"name"`

    // DeviceClassNames lists the names of all the device classes that will count against
    // the quota defined for this resource name
    // +listType=atomic
    // +kubebuilder:validation:MaxItems=8
    DeviceClassNames []corev1.ResourceName `json:"deviceClassNames"`
}
```

The ResourceFlavor spec will have a field called dynamicResource. The cluster admin has to define the list of
deviceClasses that will be represented by a resource name and can be used to define quotas in cluster queue. With this
a ClusterQueue with nominalQuota can be defined as follows:

```yaml
apiVersion: kueue.x-k8s.io/v1alpha1
kind: DynamicResourceAllocationConfig
metadata:
  name: "default"  # Fixed name - singleton CR, only one instance allowed
  namespace: "kube-system"
spec:

>   dynamicResources:

Reviewer: ```suggestion
  resources:
```
to align with the API

Author: Good catch, updated all such instances, found 4 more further down the doc that needed to be fixed.

Reviewer: thx, feel free to resolve

  resourceGroups:
  - coveredResources: ["cpu", "memory", "whole-gpus", "shared-gpus"]
    flavors:
    - name: "default-gpu-flavor"
      resources:
      - name: "cpu"
        nominalQuota: 9
      - name: "memory"
        nominalQuota: "1200Mi"
      - name: 'whole-gpus'
        nominalQuota: 2
      - name: 'shared-gpus'
        nominalQuota: 2
```

The above ClusterQueue is an example configuration of a queue, with half quota configured for single allocation of
example
GPUs, and half quota configured for GPUs that are shared by workloads. Similarly, when DRAPartitionableDevices feature
is supported in kubernetes, GPUs partitions can be represented by a single device class.

### Device Class Resolution and Ambiguity Handling

#### Device Class Mapping Ambiguity

Ambiguity occurs when a device class appears in multiple mappings within the DynamicResourceAllocationConfig CR. This
creates uncertainty at admission time about which resource name and corresponding quota should be used.

Example of ambiguous configuration:
```yaml
# Configuration allowing device class to map to multiple resource names
apiVersion: kueue.x-k8s.io/v1alpha1
kind: DynamicResourceAllocationConfig
metadata:

>   name: "cluster-device-mapping"

Reviewer: ```suggestion
  name: "default"
  namespace: kueue-system
```

   allocated to a single claim will be used against quota.

## Design Details

A new feature gate DynamicResourceAllocation will be introduced, allowing users to test it in dev environments while
making the changes dormant for production users. The following sections will explain the design in detail.

### DynamicResourceAllocationConfig API

```golang
// DynamicResourceAllocationConfig is a singleton CRD that maps resource names todevice classes
// used in ClusterQueue resource quotas. It is singleton as "default" is the only allowed named for the CRD instance in
// Kueue namespace.
type DynamicResourceAllocationConfig struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec defines the desired state of DynamicResourceAllocationConfig
    Spec DynamicResourceAllocationConfigSpec `json:"spec"`
}

// DynamicResourceAllocationConfigSpec defines the mappings between resource names and device classes
type DynamicResourceAllocationConfigSpec struct {
    // Resources is a list of mappings from resource name to device classes

>     // +listType=atomic

Reviewer 2: Why do we need to have `atomic` instead of `map`? Is this `resources` field order important?
If we can accept this field as any order, we should use the following:

```suggestion
    // +listType=map
    // +listMapKey=name
```

Reviewer: +1 for using map, yes, duplicates are not supported, the resource canonical name should be used as the key

    // +kubebuilder:validation:MaxItems=16
    Resources []DynamicResource `json:"resources"`
}

// DynamicResource defines a mapping from a resource name to a list of device classes
type DynamicResource struct {

>     // Name is the resource name that will be referred to in ClusterQueue and Workload status

Reviewer 2: ```suggestion
    // Name is the resource name that will be referred to in ClusterQueue and Workload admission status.
```

As I mentioned in another thread, could we say as `admission status`?

    Name corev1.ResourceName `json:"name"`

    // DeviceClassNames lists the names of all the device classes that will count against
    // the quota defined for this resource name

>     // +listType=atomic

Reviewer 2: Same as another comment, do we need this as an atomic? Or is this field acceptable in any order?

Reviewer 2: I'm ok with either typed, `map` / `set`. The point is if we can return the duplication errors to users when they accidentally add duplicated values.

@alaypatel07 Do you know if `set` supports raising validation error and return them to users? Or is it just output the error log in kueue-controller-manager?

Reviewer 2: > > @alaypatel07 Do you know if set supports raising validation errors and returning them to users? Or is it just outputting the error log in the kueue-controller-manager?
> 
> I assumed this is going to be a openapi validation, so client will get the message through validation before it even hits the webhook/kueue stack

Thank you for describing that. In that case, we can select `set` instead of `map` which reduce annotations.

Reviewer 2: > tbh I think like the validation can also be determined later during implementation phase, because then we will add a test, and now we rely static analysis.

It looks like the problems has been resolved. So, we can move to the next step.

Author: updated.
